streams/src/main/java/org/apache/kafka/streams/kstream/PrintForeachAction.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.kstream;
+
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serde;
+
+import java.io.PrintWriter;
+
+public class PrintForeachAction<K, V> implements ForeachAction<K, V> {
+
+    private final String streamName;
+
+    private Serde<?> keySerde;
+    private Serde<?> valueSerde;
+    
+    private final PrintWriter printWriter;
+    private ProcessorContext context;
+
+    public PrintForeachAction(final PrintWriter printWriter, final Serde<?> keySerde, final Serde<?> valueSerde, final String streamName) {
```

Comment:
I was not thinking it through when proposing this change. While reading the code now I realized there are a few more stateful members such as `init` setting serdes and closure that cannot be easily wrapped in a single-functioned stateless `ForeachAction` interface, and causing the current implementations leaking the info out of the interface in order to do that.

Trying to remedy this by proposing a slightly different approach: adding a `KStreamPrint` that extends `KStreamPeek`, so that the latter stays as pure stateless while the former adds all the stateful logic:

1. taking the passed in writer initialize the serdes upon calling `init`.
2. calling `maybeDeserialize` and passing bytes to the underlying `ForeachAction` whose types are always `<byte[], byte[]>`.
2. closing the writer upon `close`.

By doing this we will not completely get rid of `KeyValuePrinter` and hence have less LOC savings, but it may be a better code hierarchy. WDYT @jedichien ?